Frameworks/CoreGraphics/CGContext.mm

```diff
@@ -2356,14 +2360,105 @@ void CGContextDrawTiledImage(CGContextRef context, CGRect rect, CGImageRef image
 
 #pragma region Drawing Operations - Gradient + Shading
 /**
- @Status Stub
+* Insert a transparent color at the specified 'location'.
+* This will also move the color at the specified 'location' to the supplied 'position'
+*/
+static inline void __CGGradientInsertTransparentColor(std::vector<D2D1_GRADIENT_STOP>& gradientStops, int location, float position) {
+    gradientStops[location].position = position;
+    // set the edge location to be transparent
+    D2D1_GRADIENT_STOP transparent;
+    transparent.color = D2D1::ColorF(0, 0, 0, 0);
+    transparent.position = location;
+    gradientStops.push_back(transparent);
+}
+
+/*
+* Convert CGGradient to D2D1_GRADIENT_STOP
+*/
+static std::vector<D2D1_GRADIENT_STOP> __CGGradientToD2D1GradientStop(CGContextRef context,
+                                                                      CGGradientRef gradient,
+                                                                      CGGradientDrawingOptions options) {
+    unsigned long gradientCount = _CGGradientGetCount(gradient);
+    std::vector<D2D1_GRADIENT_STOP> gradientStops(gradientCount);
+
+    CGFloat* colorComponents = _CGGradientGetColorComponents(gradient);
+    CGFloat* locations = _CGGradientGetStopLocation(gradient);
+    for (unsigned long i = 0; i < gradientCount; ++i) {
+        unsigned int colorIndex = (i * 4);
+        gradientStops[i].color = D2D1::ColorF(colorComponents[colorIndex],
+                                              colorComponents[colorIndex + 1],
+                                              colorComponents[colorIndex + 2],
+                                              colorComponents[colorIndex + 3]);
+        gradientStops[i].position = locations[i];
+    }
+
+    // we want to support CGGradientDrawingOptions, but by default d2d will extend the region via repeating the brush (or other
+    // effect based on the extend mode).   We support that by inserting a point (with transparent color) close to the start/end points, such
+    // that d2d will automatically extend the transparent color, thus we obtain the desired effect for CGGradientDrawingOptions.
+
+    switch (options) {
```

Comment:
This could be modelled as two ifs. It looks like that is the intended configuration, since the two constants are bitshifted.

```c
if (!(options & kCGGradientDrawsAfterEndLocation)) {
    // insert transparent end
}

if (!(options & ...before start...)) {
    // insert transparent start
}
```

Using a switch requires you to duplicate the stops in two different branches.